flips/20220516-reference-creation-semantics.md

```diff
@@ -0,0 +1,191 @@
+means that `&v` will create a reference to the value `v`, while `*p` will obtain the value pointed-to
+by the pointer `p`. Confusingly, C uses the `*` postfix to denote a pointer type, so if `v` has an `int`
+type, then `&v` has type `int*`.
+
+Other languagues derived from C like Rust have also chosen this syntax, although in Rust's case for a
+pointer type the `*` appears at the beginning of the type instead of the end, like `*int`.
+
+### SML/OCaml/F#
+
+SML and derived languages like OCaml or F# use the `ref` keyword to create a value has a reference type, 
+as well as to denote a type that is a reference. A reference functions like a "box" that contains the 
+value inside, and the value cannot be accessed directly without first "unboxing" the reference. This 
+syntax has the benefit of being very obvious about what is and is not a reference, but is unfamiliar 
+to most developers, and also differs from the existing `&` type prefix that Cadence already uses to 
+denote a reference type. 
+
+## Design Proposal
+
+### Explanation
+
+This FLIP proposes to simply remove the type component of taking a 
+reference, and to allow people to write `&v` to take a reference like in C or C++. 
+Additionally, the `as` operator would only be the upcasting operator; it would no 
+longer have any specific relationship with referencing. This has the
+nice property that it will not require most code to be changed: any existing
+instances of code written as `&v as &T` will simply be interpreted as first
+taking the reference `&v`, and then casting that expression to `&T`, which 
+has exactly the existing behavior. 
+
+However, when references were first added to Cadence, a primary concern was that they could
+expose functionality of resources by allowing references to those resources to be distributed
+to others. The original `as` cast-like syntax for references was used to require references
+to be given an explicit type to make it less likely for developers to unintentionally 
+expose the internals of their resources. We would like to preserve this safeguard, so we will 
+also add a new check to the checker requiring that `&v` be given an explicit type.
+As an example, the following code would create a reference to a value `r`:
+
+```cadence
+resource R { /* ... */ }
+
+let r <- create R() // r has type @R
+let p = &r as &R // p has type &R
+```
+
+This code would do the same thing:
+
+```cadence
+resource R { /* ... */ }
+
+let r <- create R() // r has type @R
+let p: &R = &r // p has type &R
+```
+
+Meanwhile, this code would have a type error:
+
+```cadence
+resource R { /* ... */ }
+
+let r <- create R() // r has type @R
+let p = &r // type error: `&r` does not have a type provided
+```
+
+Authorized references can be created the same way by using the `auth` keyword as normal.
+
+```cadence
+resource R { /* ... */ }
+
+let r <- create R() // r has type @R
+let p = &r as auth &R // p has type auth &R
```

Comment:
Maybe also add an example for auth-reference where the type is coming from the context.
e.g:
```cadence
let p: auth &R = &r
```